File: app/schemas/engineering.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class EngineeringDesignCreate(BaseModel):
    project_id: int = Field(ge=1)
    well_no: str = Field(min_length=1, max_length=64)
    template_type: str = Field(default="word", pattern=r"^(word|excel)$")


class EngineeringDesignUpdate(BaseModel):
    remark: str | None = None


class EngineeringDesignQuery(BaseModel):
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=200)
    well_no: str | None = None
    project_id: int | None = None


class EngineeringDesignOut(BaseModel):
    id: int
    project_id: int | None = None
    well_no: str
    version: str
    minio_bucket: str
    minio_object_key: str
    checksum: str | None = None
    remark: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class RuleCheckResult(BaseModel):
    passed: bool
    errors: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class DesignGenerateOut(BaseModel):
    design: EngineeringDesignOut
    rule_check: RuleCheckResult

File: frontend/src/api/engineering.ts
import { http, unwrap } from './http'
import type { PageResult } from '../types/workover'

export interface EngineeringDesign {
  id: number
  project_id: number | null
  well_no: string
  version: string
  minio_bucket: string
  minio_object_key: string
  checksum?: string | null
  remark?: string | null
  created_at: string
  updated_at: string
}

export interface RuleCheckResult {
  passed: boolean
  errors: string[]
  warnings: string[]
}

export interface DesignGenerateResult {
  design: EngineeringDesign
  rule_check: RuleCheckResult
}

export function listDesigns(query: { page?: number; page_size?: number; well_no?: string; project_id?: number }) {
  return unwrap<PageResult<EngineeringDesign>>(http.get('/engineering-designs/', { params: query }))
}

export function generateDesign(payload: { project_id: number; well_no: string; template_type: 'word' | 'excel' }) {
  return unwrap<DesignGenerateResult>(http.post('/engineering-designs/generate', payload))
}

export function checkDesignRules(projectId: number) {
  return unwrap<RuleCheckResult>(http.post('/engineering-designs/check-rules', null, { params: { project_id: projectId } }))
}

export function getDesignDownloadUrl(id: number) {
  return unwrap<{ download_url: string; expire_seconds: number }>(http.get(`/engineering-designs/${id}/download`))
}

export function deleteDesign(id: number) {
  return unwrap<void>(http.delete(`/engineering-designs/${id}`))
}
